### ISSUE TITLE
Refactor the mechanism for keeping track of exit code to avoid race condition

### ISSUE THREAD

**Issue by charlesetc** [charlesetc/feather#6]
Right now there is a race condition with the following code:

```ocaml
open Feather

let () =
  process "a" [] |> run_bg;
  process "b" [] |> run;
  print_int (Feather.last_exit ())
```

It could sometimes print the exit status for process "a" if it happens to finish between process "b" exiting and the call to last_exit. Instead the interface should expose some way to get the exit status in run. (Maybe with an optional GADT parameter that allows `run` to return a different type?)

**Comment by Firobe** [comment]
Hey, I came up with a potential solution involving GADT for this, which also involves unifying `run` and `collect_stdout` (and also adds a way to collect stderr).
I implemented it on some toy code to check everything made sense and the compiler was happy but I wanted your feedback before trying to implement it on Feather (this is mostly me having fun with GADTs so feel free to reject this if it's too complicated for what it's worth)

User-wise, executing commands without collecting anything would not change. However the status would be definitely lost (`last_exit` would be removed).

```ocaml
let A = process "a" [] in
let B = process "b" [] in
A |> run;
A |> run_bg;
```
If the user wants to collect anything about the command (status, stdout, stderr), she has to use the `want` function then a function for each wanted information (in any order), between the command and `run`.
```ocaml
ignore ( A |> want |> status |> run );
ignore ( A |> want |> status |> stdout |> stderr |> run );
```
Using collecting functions without or before `want` does not type. Same for using `want` multiple times. Same for putting `want` or collecting function inside the command itself.
```ocaml
(* Does not type *)
ignore ( A |> status |> run );
ignore ( A |> status |> want |> run);
ignore ( A |> want |> want |> status |> run );
ignore ( (A |> want |> status) |. B |> run );
```
If the user collects info about the command, then `run` always return a record with three fields `stdout`, `status` and `stderr`. However the type of the fields depend on what the user selects.
```ocaml
A |> want |> status |> run ;; (* {status=42; stderr=(); stdout=()} *)
A |> want |> stdout |> run ;; (* {status=(); stderr=(); stdout="this is stdout"} *)
A |> want |> stdout |> status |> run ;; (* {status=42; stderr=(); stdout="this is stdout"} *)
```
This provides a type-safe way to gather info, while keeping things simple if we don't want to collect anything.
```ocaml
let res = A |> want |> status |> run in
print_int res.status; (* OK *)
print_string res.stdout; (* Does not type : expected string but got unit *)
```
Lastly, it's possible to define the existing specialized functions based on this approach.
```ocaml
let collect_stdout cmd =
  (cmd |> want |> stdout |> run).stdout
in
let str = A |> collect_stdout in
print_string str
```

**Comment by charlesetc** [comment]
This is really cool, thanks! I want to think a bit about this more before deciding anything but some quick thoughts:

* I am a bit hesitant to use GADTs because I think they can sometimes make things less approachable for beginners but yeah it might end up offering the best interface... I'm not sure.
* I think I would probably want to use `collect` or `capture` instead of `want`.
* I am curious if it's possible to do somthing like `collect (stdout >> stderr >> status)` using [composition_infix](https://github.com/janestreet/core_kernel/blob/master/composition_infix/src/composition_infix.mli) instead of a chain of `|>` since it might be a bit clearer that the info-getting functions are relevant to `collect`.

Anyways yeah this is exciting! The current exit code system definitely leaves something to be desired.

**Comment by Firobe** [comment]
Thanks! To address your thoughts:
- Yeah, GADTs complicates the type of the objects we're carrying around and some error messages. I'm not sure either what the best solution is.
- Definitely agree with `collect` or `capture` (maybe more `collect` since capturing the status makes a little less sense)
- It's definitely possible to do what you suggest with the composition operator (I adapted my toy example to test it) and I think it's more legible indeed

**Comment by charlesetc** [comment]
Sounds great! Agreed. Anyway yeah I might do some refactoring or editing along the way, but I'm definitely open to improvements!

**Comment by Firobe** [comment]
Sure! I'm gonna wait until you have a look at https://github.com/charlesetc/feather/pull/8 before working on implementing this into Feather.

**Comment by charlesetc** [comment]
Fixed in #10! Thanks for the huge improvement!
